app/femr/data/daos/system/PatientRepository.java

```diff
@@ -351,17 +353,37 @@ public IPatient retrievePatientById(Integer id) {
     @Override
     public IPatient savePatient(IPatient patient) {
 
-        IPatient response = null;
-        try {
-
-            Ebean.save(patient);
-        } catch (Exception ex) {
+        Optional<Patient> maxIdPatient;
+        Boolean savedSafely = false;
+        while (!(savedSafely)) {
+            try {
+                maxIdPatient = QueryProvider.getPatientQuery()
+                        .where().eq("kit_id", patient.getPatientKey().getKitId())
+                        .orderBy("id DESC")
+                        .setMaxRows(1)
+                        .findOneOrEmpty();
+
+                if (maxIdPatient.isPresent()) {
+                    patient.setId(maxIdPatient.get().getId() + 1);
+                } else {
+                    patient.setId(1);
+                }
+                Ebean.save(patient);
+                savedSafely = true;
+            } catch (io.ebean.DuplicateKeyException ex) {
+                // id must have just been taken, try again
+                System.out.println("duplicate key exception, try to get a valid id again");
```

Comment:
It's not totally an error as this case is just when two devices running on the same try to add a patient at the same time. This is basically a bottle neck to make sure each time a patient is added, each patient gets assigned a unique patient id (for the kit its on). I can change it to a log though, since that makes sense whether or not this is an error. Do we have a logger though?

Reply:
yes there is a logger